
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html 
     PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
     "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<!--

    Methods

  -->
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
  <title>Methods</title>
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
  <link rel="stylesheet" href="rdoc-style.css" type="text/css" />
  <base target="docwin" />
</head>
<body>
<div id="index">
  <h1 class="section-bar">Methods</h1>
  <div id="index-entries">
    <a href="classes/Toolbox/HomologDB/AlignmentSlice.html#M000284">_create_underlying_alignments (Toolbox::HomologDB::AlignmentSlice)</a><br />
    <a href="classes/Toolbox/NcRNA/Parser.html#M000314">_dollo_count (Toolbox::NcRNA::Parser)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000218">_intronic_at_nodes (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/HomologDB/Mapper.html#M000268">_map (Toolbox::HomologDB::Mapper)</a><br />
    <a href="classes/Toolbox/SVGWriter/Graph.html#M000193">add_element (Toolbox::SVGWriter::Graph)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000253">aligned_nucleotide_to_fasta (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000254">aligned_peptide_to_fasta (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000257">aligned_position (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000252">aligned_seq (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyAlignment.html#M000270">alignment_length (Toolbox::HomologDB::HomologyAlignment)</a><br />
    <a href="classes/Toolbox/NcRNA/Parser.html#M000310">all_xml_to_sql (Toolbox::NcRNA::Parser)</a><br />
    <a href="classes/Ensembl/Core/Transcript.html#M000046">all_xrefs (Ensembl::Core::Transcript)</a><br />
    <a href="classes/Ensembl/Core/Translation.html#M000052">all_xrefs (Ensembl::Core::Translation)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000113">belongs_to_group? (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000106">best_mirbase_match (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/String.html#M000012">binary_to_ascii (String)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000255">biograph_coordinates (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Ensembl/Core/Transcript.html#M000045">cdna_slices (Ensembl::Core::Transcript)</a><br />
    <a href="classes/Toolbox/Nexus/Writer.html#M000172">characters (Toolbox::Nexus::Writer)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000031">check_taxa (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="classes/Toolbox/ToolDB/GoTerm.html#M000164">check_terms (Toolbox::ToolDB::GoTerm)</a><br />
    <a href="classes/Bio/FastaFormat.html#M000040">cigar_line (Bio::FastaFormat)</a><br />
    <a href="classes/ActiveRecord/Base.html#M000016">clear_connections (ActiveRecord::Base)</a><br />
    <a href="classes/Toolbox/GoDB/DBConnection.html#M000208">connect (Toolbox::GoDB::DBConnection)</a><br />
    <a href="classes/Toolbox/RfamDB/DBConnection.html#M000307">connect (Toolbox::RfamDB::DBConnection)</a><br />
    <a href="classes/Toolbox/ToolDB/DBConnection.html#M000154">connect (Toolbox::ToolDB::DBConnection)</a><br />
    <a href="classes/Toolbox/MetagenomeDB/DBConnection.html#M000304">connect (Toolbox::MetagenomeDB::DBConnection)</a><br />
    <a href="classes/Toolbox/Neanderthal/DBConnection.html#M000181">connect (Toolbox::Neanderthal::DBConnection)</a><br />
    <a href="classes/Toolbox/MitoDB/DBConnection.html#M000190">connect (Toolbox::MitoDB::DBConnection)</a><br />
    <a href="classes/Toolbox/FlyDB/DBConnection.html#M000306">connect (Toolbox::FlyDB::DBConnection)</a><br />
    <a href="classes/Toolbox/HomologDB/DBConnection.html#M000296">connect (Toolbox::HomologDB::DBConnection)</a><br />
    <a href="classes/Toolbox/NcRNA/Entry.html#M000339">connect_to_genome (Toolbox::NcRNA::Entry)</a><br />
    <a href="classes/Toolbox/HomologDB/Organism.html#M000298">connect_to_genome (Toolbox::HomologDB::Organism)</a><br />
    <a href="classes/Toolbox/MitoDB/Organism.html#M000191">connect_to_genome (Toolbox::MitoDB::Organism)</a><br />
    <a href="classes/Toolbox/Converter/Tables.html#M000198">convert (Toolbox::Converter::Tables)</a><br />
    <a href="classes/Toolbox/Converter/Phylip.html#M000203">convert_names (Toolbox::Converter::Phylip)</a><br />
    <a href="classes/Array.html#M000007">coordinates_overlap? (Array)</a><br />
    <a href="classes/Toolbox/DBParser/Summary.html#M000241">create_dollo_file (Toolbox::DBParser::Summary)</a><br />
    <a href="classes/Toolbox/DBParser/Summary.html#M000238">create_mesquite_file (Toolbox::DBParser::Summary)</a><br />
    <a href="classes/Toolbox/Converter/Formats.html#M000201">create_phylip (Toolbox::Converter::Formats)</a><br />
    <a href="classes/Toolbox/Table.html#M000349">create_table (Toolbox::Table)</a><br />
    <a href="classes/Toolbox/ToolDB/GoTerm.html#M000158">description (Toolbox::ToolDB::GoTerm)</a><br />
    <a href="classes/Toolbox/ToolDB/HostgeneGoterm.html#M000152">description (Toolbox::ToolDB::HostgeneGoterm)</a><br />
    <a href="classes/Toolbox/HomologDB/InfernalHit.html#M000299">details (Toolbox::HomologDB::InfernalHit)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyGroup.html#M000280">do_align (Toolbox::HomologDB::HomologyGroup)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000145">do_align (Toolbox::ToolDB::Grouping)</a><br />
    <a href="files/lib/homology_tester_rb.html#M000001">do_align (lib/homology_tester.rb)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyAlignment.html#M000275">do_local_align (Toolbox::HomologDB::HomologyAlignment)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyAlignment.html#M000273">do_ner_align (Toolbox::HomologDB::HomologyAlignment)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000221">dollo_family_presence (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000217">dollo_nodes_stats (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000125">dollo_state_at_node (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/Family.html#M000146">dollo_state_at_node (Toolbox::ToolDB::Family)</a><br />
    <a href="classes/Toolbox/Table.html#M000347">draw_table (Toolbox::Table)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000229">duplication (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/ToolDB/Organism.html#M000166">ensembl_db_connection (Toolbox::ToolDB::Organism)</a><br />
    <a href="classes/Toolbox/Neanderthal/Blast.html#M000174">ensembl_gene (Toolbox::Neanderthal::Blast)</a><br />
    <a href="classes/Toolbox/Neanderthal/Blast.html#M000175">ensembl_seq_region (Toolbox::Neanderthal::Blast)</a><br />
    <a href="classes/Toolbox/Neanderthal/Blast.html#M000176">ensembl_start (Toolbox::Neanderthal::Blast)</a><br />
    <a href="classes/Toolbox/Neanderthal/Blast.html#M000177">ensembl_stop (Toolbox::Neanderthal::Blast)</a><br />
    <a href="classes/Toolbox/Neanderthal/Blast.html#M000178">ensembl_strand (Toolbox::Neanderthal::Blast)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000262">exon_overlaps? (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Ensembl/Compara/Member.html#M000078">exonerate_homologs (Ensembl::Compara::Member)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000224">expression_at_nodes (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000121">expression_levels (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000120">expression_stats (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000223">family_copies (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000222">family_presence (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000219">family_summary (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/Converter/Formats.html#M000200">fasta_to_phylip (Toolbox::Converter::Formats)</a><br />
    <a href="classes/Toolbox/ToolDB/TreeNode.html#M000124">fetch_all_hostgenes (Toolbox::ToolDB::TreeNode)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000111">fetch_all_ncrnas (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000094">fetch_all_targets (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000082">fetch_conserved (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000142">fetch_diff_scores (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000088">fetch_ensembl_gene (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Ensembl/Compara/Member.html#M000074">fetch_ensembl_gene (Ensembl::Compara::Member)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000115">fetch_ensembl_gene (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000141">fetch_ensembl_genomic_align_block (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000089">fetch_ensembl_member (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000116">fetch_ensembl_member (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/ToolDB/GoTerm.html#M000157">fetch_go_slims (Toolbox::ToolDB::GoTerm)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000118">fetch_goterms (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000099">fetch_group_members (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000068">fetch_hostgene (Ensembl::Core::Gene)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000095">fetch_hostgene (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000127">fetch_hostgenes (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/TreeNode.html#M000123">fetch_hostgenes (Toolbox::ToolDB::TreeNode)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000134">fetch_ncrna_group (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000133">fetch_ncrna_plain_name (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/Organism.html#M000165">fetch_ncrnas (Toolbox::ToolDB::Organism)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000112">fetch_ncrnas (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000126">fetch_ncrnas (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000266">fetch_ner_regions (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000100">fetch_node_presence (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000101">fetch_nodes (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Ensembl/Compara/GenomicAlign.html#M000071">fetch_noncoding (Ensembl::Compara::GenomicAlign)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000070">fetch_noncoding (Ensembl::Core::Gene)</a><br />
    <a href="classes/Ensembl/Compara/GenomicAlignBlock.html#M000079">fetch_noncoding (Ensembl::Compara::GenomicAlignBlock)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000140">fetch_organisms (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000265">fetch_overlapping_alignments (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyGroup.html#M000281">fetch_overlapping_alignments (Toolbox::HomologDB::HomologyGroup)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000117">fetch_pan_member (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/ToolDB/TreeNode.html#M000122">fetch_present_groups (Toolbox::ToolDB::TreeNode)</a><br />
    <a href="classes/Toolbox/ToolDB/ExpressionTissue.html#M000155">fetch_tissues (Toolbox::ToolDB::ExpressionTissue)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000137">fetch_unique_goterms (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000023">ffd_codons (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000085">find_external_id (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/Converter/Trees.html#M000207">from_phylip (Toolbox::Converter::Trees)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000143">gal_start (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000144">gal_stop (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000030">gap_score (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000059">genbank_id (Ensembl::Core::Gene)</a><br />
    <a href="classes/Toolbox/NcRNA/Selection.html#M000324">gene_alignment (Toolbox::NcRNA::Selection)</a><br />
    <a href="classes/Ensembl/Core/Xref.html#M000051">genes (Ensembl::Core::Xref)</a><br />
    <a href="classes/Toolbox/HomologDB/InfernalHit.html#M000301">genomic_start (Toolbox::HomologDB::InfernalHit)</a><br />
    <a href="classes/Toolbox/HomologDB/InfernalHit.html#M000302">genomic_stop (Toolbox::HomologDB::InfernalHit)</a><br />
    <a href="classes/Toolbox/HomologDB/InfernalHit.html#M000303">genomic_strand (Toolbox::HomologDB::InfernalHit)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000131">gerp_mean (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000132">gerp_median (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000130">gerp_scores (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000248">get_aligned_track (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Ensembl/Compara/Member.html#M000076">get_alignment (Ensembl::Compara::Member)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyAlignment.html#M000278">get_alignment_slice (Toolbox::HomologDB::HomologyAlignment)</a><br />
    <a href="classes/Toolbox/ToolDB/DolloNode.html#M000149">get_children (Toolbox::ToolDB::DolloNode)</a><br />
    <a href="classes/Toolbox/ToolDB/GoTerm.html#M000159">get_children (Toolbox::ToolDB::GoTerm)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyGroup.html#M000279">get_clustalw (Toolbox::HomologDB::HomologyGroup)</a><br />
    <a href="classes/Toolbox/MitoDB/HomologyAlignment.html#M000186">get_clustalw (Toolbox::MitoDB::HomologyAlignment)</a><br />
    <a href="classes/Toolbox/MitoDB/HomologyMember.html#M000183">get_clustalw (Toolbox::MitoDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyAlignment.html#M000272">get_clustalw (Toolbox::HomologDB::HomologyAlignment)</a><br />
    <a href="classes/Toolbox/HomologDB/AlignmentSlice.html#M000283">get_clustalw (Toolbox::HomologDB::AlignmentSlice)</a><br />
    <a href="classes/Toolbox/Table.html#M000348">get_color (Toolbox::Table)</a><br />
    <a href="classes/Toolbox/HomologDB/Member.html#M000289">get_ensembl_gene (Toolbox::HomologDB::Member)</a><br />
    <a href="classes/Ensembl/PanCompara/Member.html#M000081">get_ensembl_gene (Ensembl::PanCompara::Member)</a><br />
    <a href="classes/Ensembl/Core/Translation.html#M000053">get_externaldb_accs (Ensembl::Core::Translation)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000063">get_externaldb_accs (Ensembl::Core::Gene)</a><br />
    <a href="classes/Ensembl/Core/Transcript.html#M000047">get_externaldb_accs (Ensembl::Core::Transcript)</a><br />
    <a href="classes/Toolbox/NcRNA/ContigMap.html#M000320">get_features (Toolbox::NcRNA::ContigMap)</a><br />
    <a href="classes/Toolbox/MitoDB/Member.html#M000187">get_gene (Toolbox::MitoDB::Member)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000138">get_genomic_align_block (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000139">get_genomic_align_block_id (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/HomologDB/Member.html#M000286">get_genomic_region (Toolbox::HomologDB::Member)</a><br />
    <a href="classes/Ensembl/Compara/Member.html#M000077">get_homology_seqs (Ensembl::Compara::Member)</a><br />
    <a href="classes/Ensembl/PanCompara/Member.html#M000080">get_longest_peptide_sequence (Ensembl::PanCompara::Member)</a><br />
    <a href="classes/Ensembl/Compara/Member.html#M000073">get_longest_peptide_sequence (Ensembl::Compara::Member)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000069">get_longest_transcript (Ensembl::Core::Gene)</a><br />
    <a href="classes/Toolbox/HomologDB/Member.html#M000287">get_longest_transcript (Toolbox::HomologDB::Member)</a><br />
    <a href="classes/Toolbox/HomologDB/AlignmentSlice.html#M000285">get_mapper (Toolbox::HomologDB::AlignmentSlice)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyAlignment.html#M000277">get_mapper (Toolbox::HomologDB::HomologyAlignment)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyAlignment.html#M000274">get_ner_coords (Toolbox::HomologDB::HomologyAlignment)</a><br />
    <a href="classes/Toolbox/HomologDB/Member.html#M000295">get_pan_member (Toolbox::HomologDB::Member)</a><br />
    <a href="classes/Toolbox/ToolDB/DolloNode.html#M000148">get_parent (Toolbox::ToolDB::DolloNode)</a><br />
    <a href="classes/Toolbox/ToolDB/GoTerm.html#M000160">get_parents (Toolbox::ToolDB::GoTerm)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000065">get_pfam_features (Ensembl::Core::Gene)</a><br />
    <a href="classes/Toolbox/MitoDB/Member.html#M000188">get_protein (Toolbox::MitoDB::Member)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000260">get_rnaz (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Ensembl/Core/Transcript.html#M000048">get_snorna_region (Ensembl::Core::Transcript)</a><br />
    <a href="classes/Toolbox/HomologDB/Member.html#M000294">get_snornas (Toolbox::HomologDB::Member)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000264">get_snornas (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Bio/Tree.html#M000039">get_taxa (Bio::Tree)</a><br />
    <a href="classes/Toolbox/MitoDB/Member.html#M000189">get_transcript (Toolbox::MitoDB::Member)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000227">go_hierarchy (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/DBParser/Summary.html#M000239">go_stat (Toolbox::DBParser::Summary)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000233">go_stat (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000225">go_terms_at_nodes (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000067">go_terms_with_evidence (Ensembl::Core::Gene)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000098">group (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/ToolDB/GoTerm.html#M000161">has_children? (Toolbox::ToolDB::GoTerm)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000097">has_group? (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000104">has_hostgene? (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000114">has_ncrna? (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/ToolDB/GoTerm.html#M000162">has_parents? (Toolbox::ToolDB::GoTerm)</a><br />
    <a href="classes/Toolbox/Nexus/Writer.html#M000170">header (Toolbox::Nexus::Writer)</a><br />
    <a href="classes/Toolbox/ToolDB/GoTerm.html#M000163">hierarchy (Toolbox::ToolDB::GoTerm)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000220">hostgene_at_nodes (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000231">hostgene_content (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/HomologDB/NcrnaMember.html#M000246">hostgene_start (Toolbox::HomologDB::NcrnaMember)</a><br />
    <a href="classes/Toolbox/HomologDB/NcrnaMember.html#M000247">hostgene_stop (Toolbox::HomologDB::NcrnaMember)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000108">hosts_ncrna? (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/DBParser/Summary.html#M000240">html_gene_list (Toolbox::DBParser::Summary)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000250">infernal_hits (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/NcRNA/ContigMap.html#M000319">initalize (Toolbox::NcRNA::ContigMap)</a><br />
    <a href="classes/Toolbox/Converter/Phylip.html#M000202">intialize (Toolbox::Converter::Phylip)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000213">intronic_groups (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyAlignment.html#M000276">is_aligned? (Toolbox::HomologDB::HomologyAlignment)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000066">is_intronic? (Ensembl::Core::Gene)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000103">is_intronic? (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000128">is_intronic? (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000087">is_sense? (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000129">is_sense? (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/GoTerm.html#M000156">is_slim? (Toolbox::ToolDB::GoTerm)</a><br />
    <a href="classes/Toolbox/ToolDB/HostgeneGoterm.html#M000150">is_slim? (Toolbox::ToolDB::HostgeneGoterm)</a><br />
    <a href="classes/Toolbox/Neanderthal/Blast.html#M000179">length (Toolbox::Neanderthal::Blast)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000232">list_by_rfam (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/NcRNA/MIRBASE.html#M000336">map_family (Toolbox::NcRNA::MIRBASE)</a><br />
    <a href="classes/Toolbox/NcRNA/Parser.html#M000311">map_mirnas (Toolbox::NcRNA::Parser)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000256">map_peptide_position (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000032">map_position (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="classes/Ensembl/Core/Transcript.html#M000049">map_slice (Ensembl::Core::Transcript)</a><br />
    <a href="classes/Ensembl/Compara/GenomicAlign.html#M000072">mark_feature (Ensembl::Compara::GenomicAlign)</a><br />
    <a href="classes/String.html#M000015">mask (String)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000026">mask_and_strip (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="files/lib/extensions_rb.html#M000002">mask_seq (lib/extensions.rb)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000062">mirbase (Ensembl::Core::Gene)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000091">mirbase_acc (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000093">mirbase_family (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000230">mirbase_targets_and_nodes (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/ToolDB/HostgeneGoterm.html#M000151">name (Toolbox::ToolDB::HostgeneGoterm)</a><br />
    <a href="classes/Toolbox/ToolDB/HostgeneGoterm.html#M000153">namespace (Toolbox::ToolDB::HostgeneGoterm)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000086">naseq (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000061">ncrna_id (Ensembl::Core::Gene)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000261">ncrna_overlaps? (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000215">ncrnas_per_node (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000214">ncrnas_per_node_dollo (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/SVGWriter/Graph.html#M000192">new (Toolbox::SVGWriter::Graph)</a><br />
    <a href="classes/Toolbox/Converter/Formats.html#M000199">new (Toolbox::Converter::Formats)</a><br />
    <a href="classes/Toolbox/NcRNA/Blocks.html#M000333">new (Toolbox::NcRNA::Blocks)</a><br />
    <a href="classes/Bio/Vienna/RNAz.html#M000037">new (Bio::Vienna::RNAz)</a><br />
    <a href="classes/Toolbox/NcRNA/Entry.html#M000338">new (Toolbox::NcRNA::Entry)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000209">new (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/NcRNA/ConservationEntry.html#M000321">new (Toolbox::NcRNA::ConservationEntry)</a><br />
    <a href="classes/PairwiseCount.html#M000008">new (PairwiseCount)</a><br />
    <a href="classes/Toolbox/NcRNA/Statistics.html#M000330">new (Toolbox::NcRNA::Statistics)</a><br />
    <a href="classes/Bio/Exonerate.html#M000041">new (Bio::Exonerate)</a><br />
    <a href="classes/Toolbox/NcRNA/Parser.html#M000308">new (Toolbox::NcRNA::Parser)</a><br />
    <a href="classes/Toolbox/NcRNA/MIRBASE.html#M000335">new (Toolbox::NcRNA::MIRBASE)</a><br />
    <a href="classes/Toolbox/Converter/Trees.html#M000204">new (Toolbox::Converter::Trees)</a><br />
    <a href="classes/Toolbox/Nexus/Writer.html#M000168">new (Toolbox::Nexus::Writer)</a><br />
    <a href="classes/Toolbox/HomologDB/AlignmentSlice.html#M000282">new (Toolbox::HomologDB::AlignmentSlice)</a><br />
    <a href="classes/Toolbox/Table.html#M000346">new (Toolbox::Table)</a><br />
    <a href="classes/Bio/Vienna/RNAfold.html#M000035">new (Bio::Vienna::RNAfold)</a><br />
    <a href="classes/Bio/Vienna/RNAalifold.html#M000033">new (Bio::Vienna::RNAalifold)</a><br />
    <a href="classes/Toolbox/HomologDB/InfernalGroup.html#M000297">new (Toolbox::HomologDB::InfernalGroup)</a><br />
    <a href="classes/Toolbox/DBParser/Summary.html#M000234">new (Toolbox::DBParser::Summary)</a><br />
    <a href="classes/Toolbox/NcRNA/Constraints.html#M000316">new (Toolbox::NcRNA::Constraints)</a><br />
    <a href="classes/Toolbox/Parser.html#M000345">new (Toolbox::Parser)</a><br />
    <a href="classes/Toolbox/NcRNA/ConservationGroup.html#M000318">new (Toolbox::NcRNA::ConservationGroup)</a><br />
    <a href="classes/Toolbox/NcRNA/FindGroups.html#M000326">new (Toolbox::NcRNA::FindGroups)</a><br />
    <a href="classes/Toolbox/SQLinject.html#M000340">new (Toolbox::SQLinject)</a><br />
    <a href="classes/Toolbox/SVGWriter/Element.html#M000195">new (Toolbox::SVGWriter::Element)</a><br />
    <a href="classes/Bio/Infernal/Cmsearch.html#M000017">new (Bio::Infernal::Cmsearch)</a><br />
    <a href="classes/Toolbox/DBParser/Process.html#M000242">new (Toolbox::DBParser::Process)</a><br />
    <a href="classes/Toolbox/NcRNA/ConstraintGroup.html#M000325">new (Toolbox::NcRNA::ConstraintGroup)</a><br />
    <a href="classes/Toolbox/HomologDB/Mapper.html#M000267">new (Toolbox::HomologDB::Mapper)</a><br />
    <a href="classes/Toolbox/NcRNA/Selection.html#M000322">new (Toolbox::NcRNA::Selection)</a><br />
    <a href="classes/Toolbox/Converter/Tables.html#M000197">new (Toolbox::Converter::Tables)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000226">node_diff (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000135">node_values (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/NcrnaGroup.html#M000107">nodes (Toolbox::ToolDB::NcrnaGroup)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000216">nodes_expression (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/HomologDB/Member.html#M000293">nucleotide_seq (Toolbox::HomologDB::Member)</a><br />
    <a href="classes/Toolbox/HomologDB/Member.html#M000291">nucleotide_to_fasta (Toolbox::HomologDB::Member)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000259">nucleotide_to_fasta (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000249">organism_name (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/NcRNA/Statistics.html#M000331">output (Toolbox::NcRNA::Statistics)</a><br />
    <a href="classes/Toolbox/NcRNA/FindGroups.html#M000327">output (Toolbox::NcRNA::FindGroups)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000102">overlaps_exon? (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/DBParser/Summary.html#M000236">pairwise (Toolbox::DBParser::Summary)</a><br />
    <a href="classes/Bio/Exonerate.html#M000043">parse (Bio::Exonerate)</a><br />
    <a href="classes/Bio/Infernal/Cmsearch.html#M000019">parse (Bio::Infernal::Cmsearch)</a><br />
    <a href="files/lib/extensions_rb.html#M000003">parse_baseml (lib/extensions.rb)</a><br />
    <a href="classes/Toolbox/NcRNA/Parser.html#M000313">parse_dollo_output_do_count (Toolbox::NcRNA::Parser)</a><br />
    <a href="classes/Toolbox/NcRNA/Parser.html#M000315">parse_dollo_output_for_sql (Toolbox::NcRNA::Parser)</a><br />
    <a href="classes/Toolbox/NcRNA/Parser.html#M000312">parse_family_presence (Toolbox::NcRNA::Parser)</a><br />
    <a href="classes/Toolbox/SQLinject.html#M000343">parse_full (Toolbox::SQLinject)</a><br />
    <a href="classes/Toolbox/SQLinject.html#M000342">parse_mesquite (Toolbox::SQLinject)</a><br />
    <a href="classes/Toolbox/SQLinject.html#M000341">parse_xml (Toolbox::SQLinject)</a><br />
    <a href="classes/Ensembl/Core/Transcript.html#M000050">peptide_position (Ensembl::Core::Transcript)</a><br />
    <a href="classes/Toolbox/HomologDB/Member.html#M000292">peptide_seq (Toolbox::HomologDB::Member)</a><br />
    <a href="classes/Toolbox/HomologDB/Member.html#M000290">peptide_to_fasta (Toolbox::HomologDB::Member)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000258">peptide_to_fasta (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000027">percent_similarity (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000028">percent_similarity_by_column (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="classes/Toolbox/Converter/Trees.html#M000206">phylip (Toolbox::Converter::Trees)</a><br />
    <a href="classes/Toolbox/ToolDB/Organism.html#M000167">phylip_name (Toolbox::ToolDB::Organism)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000096">present_at? (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/HomologDB/NcrnaMember.html#M000244">present_at? (Toolbox::HomologDB::NcrnaMember)</a><br />
    <a href="classes/Toolbox/ToolDB/Grouping.html#M000136">present_at? (Toolbox::ToolDB::Grouping)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000110">present_at? (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/ToolDB/Family.html#M000147">present_at? (Toolbox::ToolDB::Family)</a><br />
    <a href="classes/Toolbox/NcRNA/FindGroups.html#M000328">print_alignment (Toolbox::NcRNA::FindGroups)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000022">print_for_paml (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000029">print_nexus (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000119">probe (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyAlignment.html#M000271">query (Toolbox::HomologDB::HomologyAlignment)</a><br />
    <a href="classes/Toolbox/MitoDB/HomologyAlignment.html#M000185">query (Toolbox::MitoDB::HomologyAlignment)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000263">region_is_aligned? (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000228">regulation_hypothesis (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000025">reverse_alignment (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000092">rfam_description (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Bio/Vienna/RNAalifold.html#M000034">run (Bio::Vienna::RNAalifold)</a><br />
    <a href="classes/Bio/Infernal/Cmsearch.html#M000018">run (Bio::Infernal::Cmsearch)</a><br />
    <a href="classes/Bio/Exonerate.html#M000042">run (Bio::Exonerate)</a><br />
    <a href="classes/Bio/Vienna/RNAfold.html#M000036">run (Bio::Vienna::RNAfold)</a><br />
    <a href="classes/Bio/Vienna/RNAz.html#M000038">run (Bio::Vienna::RNAz)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000210">run_all (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000020">run_baseml (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="classes/Toolbox/SQLinject.html#M000344">run_gerp_scores (Toolbox::SQLinject)</a><br />
    <a href="files/lib/homology_rake_file_rb.html#M000006">run_infernal (lib/homology_rake_file.rb)</a><br />
    <a href="files/lib/homology_rake_file_rb.html#M000004">run_pecan (lib/homology_rake_file.rb)</a><br />
    <a href="files/lib/homology_rake_file_rb.html#M000005">run_probcons (lib/homology_rake_file.rb)</a><br />
    <a href="classes/Toolbox/NcRNA/Statistics.html#M000332">run_search (Toolbox::NcRNA::Statistics)</a><br />
    <a href="classes/Toolbox/NcRNA/Constraints.html#M000317">run_search (Toolbox::NcRNA::Constraints)</a><br />
    <a href="classes/Toolbox/NcRNA/Blocks.html#M000334">run_search (Toolbox::NcRNA::Blocks)</a><br />
    <a href="classes/Toolbox/NcRNA/FindGroups.html#M000329">run_search (Toolbox::NcRNA::FindGroups)</a><br />
    <a href="classes/Toolbox/NcRNA/Selection.html#M000323">run_search (Toolbox::NcRNA::Selection)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000021">select_taxa (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000084">seq (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/HomologDB/NcrnaMember.html#M000243">seq (Toolbox::HomologDB::NcrnaMember)</a><br />
    <a href="classes/Toolbox/FlyDB/Transcript.html#M000305">seq (Toolbox::FlyDB::Transcript)</a><br />
    <a href="classes/Ensembl/Core/Translation.html#M000054">snorna_locations (Ensembl::Core::Translation)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000057">snornas (Ensembl::Core::Gene)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000058">softmasked_seq (Ensembl::Core::Gene)</a><br />
    <a href="classes/String.html#M000014">softmasked_seq_to_html (String)</a><br />
    <a href="classes/String.html#M000010">sql_compliant (String)</a><br />
    <a href="classes/Toolbox/MitoDB/HomologyMember.html#M000182">stable_id (Toolbox::MitoDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyMember.html#M000251">stable_id (Toolbox::HomologDB::HomologyMember)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000083">strand (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblHostgene.html#M000109">strand (Toolbox::ToolDB::EnsemblHostgene)</a><br />
    <a href="classes/Bio/Alignment/OriginalAlignment.html#M000024">strip_columns (Bio::Alignment::OriginalAlignment)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000056">structure_to_html (Ensembl::Core::Gene)</a><br />
    <a href="classes/Toolbox/DBParser/Summary.html#M000235">summarize (Toolbox::DBParser::Summary)</a><br />
    <a href="classes/Toolbox/DBParser/Summary.html#M000237">summarize_detailed (Toolbox::DBParser::Summary)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000211">summarize_nodes (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/DBParser/Ancestral.html#M000212">summarize_nodes_intronic (Toolbox::DBParser::Ancestral)</a><br />
    <a href="classes/Toolbox/MitoDB/HomologyAlignment.html#M000184">target (Toolbox::MitoDB::HomologyAlignment)</a><br />
    <a href="classes/Toolbox/HomologDB/HomologyAlignment.html#M000269">target (Toolbox::HomologDB::HomologyAlignment)</a><br />
    <a href="classes/Toolbox/NcRNA/MIRBASE.html#M000337">targets (Toolbox::NcRNA::MIRBASE)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000105">targets_hostgene? (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Toolbox/Nexus/Writer.html#M000171">taxon_list (Toolbox::Nexus::Writer)</a><br />
    <a href="classes/Toolbox/HomologDB/NcrnaMember.html#M000245">to_fasta (Toolbox::HomologDB::NcrnaMember)</a><br />
    <a href="classes/Toolbox/HomologDB/InfernalHit.html#M000300">to_fasta (Toolbox::HomologDB::InfernalHit)</a><br />
    <a href="classes/Toolbox/HomologDB/Member.html#M000288">to_fasta (Toolbox::HomologDB::Member)</a><br />
    <a href="classes/Toolbox/ToolDB/EnsemblNcrna.html#M000090">to_fasta (Toolbox::ToolDB::EnsemblNcrna)</a><br />
    <a href="classes/Ensembl/Core/Translation.html#M000055">to_fasta (Ensembl::Core::Translation)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000064">to_fasta (Ensembl::Core::Gene)</a><br />
    <a href="classes/Ensembl/Compara/Member.html#M000075">to_fasta (Ensembl::Compara::Member)</a><br />
    <a href="classes/Toolbox/Neanderthal/Contig.html#M000180">to_fasta (Toolbox::Neanderthal::Contig)</a><br />
    <a href="classes/Toolbox/Converter/Trees.html#M000205">to_nexus (Toolbox::Converter::Trees)</a><br />
    <a href="classes/Toolbox/Nexus/Writer.html#M000169">to_s (Toolbox::Nexus::Writer)</a><br />
    <a href="classes/String.html#M000011">to_snake_case (String)</a><br />
    <a href="classes/String.html#M000013">to_width (String)</a><br />
    <a href="classes/String.html#M000009">transform_date (String)</a><br />
    <a href="classes/Ensembl/Core/Transcript.html#M000044">translation_exons (Ensembl::Core::Transcript)</a><br />
    <a href="classes/Toolbox/Nexus/Writer.html#M000173">tree (Toolbox::Nexus::Writer)</a><br />
    <a href="classes/Ensembl/Core/Gene.html#M000060">uniprot_id (Ensembl::Core::Gene)</a><br />
    <a href="classes/Toolbox/SVGWriter/Element.html#M000196">write (Toolbox::SVGWriter::Element)</a><br />
    <a href="classes/Toolbox/SVGWriter/Graph.html#M000194">write (Toolbox::SVGWriter::Graph)</a><br />
    <a href="classes/Toolbox/NcRNA/Parser.html#M000309">xml_to_sql (Toolbox::NcRNA::Parser)</a><br />
  </div>
</div>
</body>
</html>